File: type-bridge-core/crates/schema-codegen/tests/acceptance/authority_rejection_check.py
from __future__ import annotations

import ast
import hashlib
import json
import shutil
import struct
import subprocess
import sys
from collections.abc import Callable
from pathlib import Path

STAGE = Path(__file__).resolve().parent
SOURCE = STAGE / "generated_ordered"
FOREIGN = STAGE / "generated_variant"
MAX_SCHEMA_AUTHORITY_BYTES = 16 * 1024 * 1024
PREFIX = "SCHEMA_AUTHORITY_BYTES: _Final[bytes] = "
SUFFIX = '.encode("utf-8")'


def canonical(value: object) -> bytes:
    return json.dumps(value, sort_keys=True, separators=(",", ":")).encode()


def envelope(package: Path) -> str:
    line = next(
        line
        for line in (package / "_authority.py").read_text().splitlines()
        if line.startswith(PREFIX)
    )
    expression = line.removeprefix(PREFIX).removesuffix(SUFFIX)
    value = ast.literal_eval(expression)
    if not isinstance(value, str):
        raise AssertionError("generated authority constant is not a string")
    return value


def fingerprint(content: object) -> str:
    digest = hashlib.sha256()
    digest.update(b"typebridge.fingerprint/v1\0")
    for value in (
        b"typebridge.schema.authority",
        b"typebridge.schema-authority/v1",
    ):
        digest.update(struct.pack(">Q", len(value)))
        digest.update(value)
    digest.update(b"\0")
    payload = canonical(content)
    digest.update(struct.pack(">Q", len(payload)))
    digest.update(payload)
    return digest.hexdigest()


def resign(value: dict[str, object]) -> None:
    authority_fingerprint = value["authority_fingerprint"]
    if not isinstance(authority_fingerprint, dict):
        raise AssertionError("authority fingerprint is not an object")
    authority_fingerprint["digest"] = fingerprint(value["content"])


def mutated(change: Callable[[dict[str, object]], None], *, resign_after: bool) -> str:
    value = json.loads(envelope(SOURCE))
    change(value)
    if resign_after:
        resign(value)
    return canonical(value).decode()


EXPECTED_DIAGNOSTIC = {
    "category": "integrity",
    "sdk_category": "integrity",
    "code": "projection_evidence_mismatch",
    "message": "Generated projection evidence does not match the verified schema package",
    "path": [{"kind": "argument", "value": "projection_evidence"}],
}


def reject(name: str, authority: str) -> None:
    package_name = f"generated_rejected_{name}"
    package = STAGE / package_name
    shutil.rmtree(package, ignore_errors=True)
    shutil.copytree(SOURCE, package)
    (package / "_authority.py").write_text(
        "from __future__ import annotations\n\n"
        "from typing import Final as _Final\n\n"
        f"{PREFIX}{authority!r}{SUFFIX}\n"
    )
    probe = (
        "import importlib, json\n"
        "try:\n"
        f"    importlib.import_module({package_name!r})\n"
        "except BaseException as error:\n"
        "    print(json.dumps({\n"
        "        'category': error.category,\n"
        "        'sdk_category': error.sdk_category,\n"
        "        'code': error.code,\n"
        "        'message': error.message,\n"
        "        'path': error.path,\n"
        "    }, sort_keys=True))\n"
        "else:\n"
        "    raise AssertionError('hostile ordered package import succeeded')\n"
    )
    completed = subprocess.run(
        [sys.executable, "-c", probe],
        cwd=STAGE,
        capture_output=True,
        text=True,
        check=False,
    )
    try:
        diagnostic = json.loads(completed.stdout)
    except json.JSONDecodeError:
        diagnostic = None
    if completed.returncode != 0 or diagnostic != EXPECTED_DIAGNOSTIC:
        raise AssertionError(
            f"{name} authority returned {completed.returncode}, "
            f"expected {EXPECTED_DIAGNOSTIC!r}, received {diagnostic!r}\n"
            f"stdout:\n{completed.stdout}\nstderr:\n{completed.stderr}"
        )


ordered_models = (SOURCE / "_models.py").read_text()
native_install = ordered_models.index("\n_install_runtime_projection(\n")
if "from ._query import " in ordered_models[:native_install]:
    raise AssertionError("ordered models imported query authority before native admission")
query_aliases = ordered_models.find("\nfrom ._query import ", native_install)
if query_aliases != -1 and query_aliases <= native_install:
    raise AssertionError("ordered model query aliases were not deferred past native admission")

ordered_init = (SOURCE / "__init__.py").read_text()
if ordered_init.index("from ._models import ") >= ordered_init.index("from ._query import "):
    raise AssertionError("ordered package init did not trigger model admission before query")
if (SOURCE / "_query.py").read_bytes() != (STAGE / "generated_v2" / "_query.py").read_bytes():
    raise AssertionError("ordered package changed the fixed Python query resource")

completed = subprocess.run(
    [sys.executable, "-c", "import generated_ordered"],
    cwd=STAGE,
    capture_output=True,
    text=True,
    check=False,
)
if completed.returncode != 0:
    raise AssertionError(
        "exact ordered generated package import failed\n"
        f"stdout:\n{completed.stdout}\nstderr:\n{completed.stderr}"
    )

reject("malformed", "{")
reject("foreign", envelope(FOREIGN))
reject(
    "stale",
    mutated(
        lambda value: value["content"]["declared_identity"].__setitem__("digest", "0" * 64),
        resign_after=False,
    ),
)
reject(
    "missing_fingerprint",
    mutated(lambda value: value.pop("authority_fingerprint"), resign_after=False),
)
reject(
    "managed_state",
    mutated(
        lambda value: value["content"]["managed_state"]["managed_semantic_schema"].__setitem__(
            "digest", "0" * 64
        ),
        resign_after=True,
    ),
)
reject(
    "capability",
    mutated(
        lambda value: (
            value["content"]["required_capabilities"].append("query.future-feature"),
            value["content"]["required_capabilities"].sort(),
        ),
        resign_after=True,
    ),
)
reject(
    "version",
    mutated(
        lambda value: value["content"].__setitem__(
            "authority_version", "typebridge.schema-authority/v2"
        ),
        resign_after=True,
    ),
)
reject(
    "oversize",
    " " * (MAX_SCHEMA_AUTHORITY_BYTES + 1),
)

print("generated Python authority rejection acceptance passed")

File: type-bridge-core/crates/schema-codegen/tests/typescript_acceptance/authority_rejection_check.mjs
import assert from "node:assert/strict";
import crypto from "node:crypto";
import { cpSync, readFileSync, rmSync, writeFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { spawnSync } from "node:child_process";
import { fileURLToPath } from "node:url";


const STAGE = dirname(fileURLToPath(import.meta.url));
const SOURCE = resolve(STAGE, "generated_v2");
const ORDERED = resolve(STAGE, "generated_ordered");
const FOREIGN = resolve(STAGE, "generated_foreign");
const MAX_SCHEMA_AUTHORITY_BYTES = 16 * 1024 * 1024;
const AUTHORITY_PREFIX = "export const SCHEMA_AUTHORITY_JSON = ";
const PROJECTION_PREFIX = "export const RUNTIME_PROJECTION_JSON = ";
const SEMANTIC_FINGERPRINT_PREFIX =
  "export const SEMANTIC_SCHEMA_FINGERPRINT_JSON = ";
const PROJECTION_FINGERPRINT_PREFIX =
  "export const PROJECTION_FINGERPRINT_JSON = ";

function sorted(value) {
  if (Array.isArray(value)) {
    return value.map(sorted);
  }
  if (value !== null && typeof value === "object") {
    return Object.fromEntries(
      Object.entries(value)
        .sort(([left], [right]) => left.localeCompare(right))
        .map(([key, child]) => [key, sorted(child)]),
    );
  }
  return value;
}

function canonical(value) {
  return JSON.stringify(sorted(value));
}

function exported(packageDirectory, relativePath, prefix) {
  const line = readFileSync(resolve(packageDirectory, relativePath), "utf8")
    .split("\n")
    .find((candidate) => candidate.startsWith(prefix));
  assert(line);
  return JSON.parse(line.slice(prefix.length, -1));
}

function replaceExport(packageDirectory, relativePath, prefix, value) {
  const path = resolve(packageDirectory, relativePath);
  const lines = readFileSync(path, "utf8").split("\n");
  const index = lines.findIndex((candidate) => candidate.startsWith(prefix));
  assert.notEqual(index, -1);
  lines[index] = `${prefix}${JSON.stringify(value)};`;
  writeFileSync(path, lines.join("\n"));
}

function envelope(packageDirectory) {
  return exported(packageDirectory, "dist/authority.js", AUTHORITY_PREFIX);
}

function projection(packageDirectory) {
  return JSON.parse(
    exported(packageDirectory, "dist/schema.js", PROJECTION_PREFIX),
  );
}

function replaceProjection(packageDirectory, value) {
  replaceExport(
    packageDirectory,
    "dist/schema.js",
    PROJECTION_PREFIX,
    typeof value === "string" ? value : canonical(value),
  );
}

function field(digest, value) {
  const encoded = Buffer.from(value);
  const length = Buffer.alloc(8);
  length.writeBigUInt64BE(BigInt(encoded.length));
  digest.update(length);
  digest.update(encoded);
}

function fingerprint(content) {
  const digest = crypto.createHash("sha256");
  digest.update(Buffer.from("typebridge.fingerprint/v1\0"));
  field(digest, "typebridge.schema.authority");
  field(digest, "typebridge.schema-authority/v1");
  digest.update(Buffer.from([0]));
  field(digest, canonical(content));
  return digest.digest("hex");
}

function resign(value) {
  value.authority_fingerprint.digest = fingerprint(value.content);
}

function mutated(change, resignAfter, packageDirectory = SOURCE) {
  const value = JSON.parse(envelope(packageDirectory));
  change(value);
  if (resignAfter) {
    resign(value);
  }
  return canonical(value);
}

function rejectLegacy(name, authority, expected) {
  const packageName = `generated_rejected_${name}`;
  const packageDirectory = resolve(STAGE, packageName);
  rmSync(packageDirectory, { recursive: true, force: true });
  cpSync(SOURCE, packageDirectory, { recursive: true });
  writeFileSync(
    resolve(packageDirectory, "dist/authority.js"),
    `// Tampered generated-authority acceptance fixture.\n\n${AUTHORITY_PREFIX}${JSON.stringify(authority)};\n`,
  );
  const completed = spawnSync(
    process.execPath,
    ["--input-type=module", "-e", `import('./${packageName}/dist/index.js')`],
    { cwd: STAGE, encoding: "utf8" },
  );
  const output = `${completed.stdout}${completed.stderr}`;
  assert.notEqual(completed.status, 0, `${name} authority unexpectedly installed`);
  assert.match(output, new RegExp(expected), `${name} authority emitted:\n${output}`);
}

const PROJECTION_EVIDENCE_MISMATCH = {
  category: "integrity",
  sdkCategory: "integrity",
  queryCategory: null,
  code: "projection_evidence_mismatch",
  message: "Generated projection evidence does not match the verified schema package",
  path: [{ kind: "argument", value: "projection_evidence" }],
  details: {},
};

function rejectSuccessor(name, tamper) {
  const packageName = `generated_projection_rejected_${name}`;
  const packageDirectory = resolve(STAGE, packageName);
  rmSync(packageDirectory, { recursive: true, force: true });
  cpSync(ORDERED, packageDirectory, { recursive: true });
  tamper(packageDirectory);
  const completed = spawnSync(
    process.execPath,
    [
      "--input-type=module",
      "-e",
      `import('./${packageName}/dist/index.js').then(() => process.exit(0), (error) => { process.stdout.write(error instanceof Error ? error.message : String(error)); process.exit(1); })`,
    ],
    { cwd: STAGE, encoding: "utf8" },
  );
  assert.notEqual(completed.status, 0, `${name} evidence unexpectedly installed`);
  assert.equal(completed.stderr, "", `${name} emitted stderr:\n${completed.stderr}`);
  assert.deepEqual(
    JSON.parse(completed.stdout),
    PROJECTION_EVIDENCE_MISMATCH,
    `${name} evidence emitted:\n${completed.stdout}`,
  );
}

rejectLegacy("malformed", "{", "malformed_canonical_json");
rejectLegacy(
  "foreign",
  envelope(FOREIGN),
  "generated_schema_authority_semantic_mismatch",
);
rejectLegacy(
  "stale",
  mutated((value) => {
    value.content.declared_identity.digest = "0".repeat(64);
  }, false),
  "generated_schema_authority_integrity_mismatch",
);
rejectLegacy(
  "missing_fingerprint",
  mutated((value) => {
    delete value.authority_fingerprint;
  }, false),
  "invalid_canonical_value",
);
rejectLegacy(
  "managed_state",
  mutated((value) => {
    value.content.managed_state.managed_semantic_schema.digest = "0".repeat(64);
  }, true),
  "generated_schema_authority_integrity_mismatch",
);
rejectLegacy(
  "capability",
  mutated((value) => {
    value.content.required_capabilities.push("query.future-feature");
    value.content.required_capabilities.sort();
  }, true),
  "unsupported_required_capability",
);
rejectLegacy(
  "version",
  mutated((value) => {
    value.content.authority_version = "typebridge.schema-authority/v2";
  }, true),
  "generated_schema_authority_unsupported_version",
);
rejectLegacy(
  "oversize",
  " ".repeat(MAX_SCHEMA_AUTHORITY_BYTES + 1),
  "canonical_json_too_large",
);

rejectSuccessor("missing_authority", (directory) => {
  const path = resolve(directory, "dist/index.js");
  const source = readFileSync(path, "utf8");
  const changed = source.replace(
    ", SCHEMA_AUTHORITY_JSON);",
    ", undefined);",
  );
  assert.notEqual(changed, source);
  writeFileSync(path, changed);
});
rejectSuccessor("malformed_authority", (directory) => {
  replaceExport(directory, "dist/authority.js", AUTHORITY_PREFIX, "{");
});
rejectSuccessor("foreign_authority", (directory) => {
  replaceExport(
    directory,
    "dist/authority.js",
    AUTHORITY_PREFIX,
    envelope(FOREIGN),
  );
});
rejectSuccessor("stale_authority", (directory) => {
  replaceExport(
    directory,
    "dist/authority.js",
    AUTHORITY_PREFIX,
    mutated((value) => {
      value.content.declared_identity.digest = "0".repeat(64);
    }, false, ORDERED),
  );
});
rejectSuccessor("unsupported_authority_version", (directory) => {
  replaceExport(
    directory,
    "dist/authority.js",
    AUTHORITY_PREFIX,
    mutated((value) => {
      value.content.authority_version = "typebridge.schema-authority/v2";
    }, true, ORDERED),
  );
});
rejectSuccessor("malformed_projection", (directory) => {
  replaceProjection(directory, "{");
});
rejectSuccessor("missing_resource", (directory) => {
  const value = projection(directory);
  value.code_resources.pop();
  replaceProjection(directory, value);
});
rejectSuccessor("extra_resource", (directory) => {
  const value = projection(directory);
  const extra = structuredClone(value.code_resources.at(-1));
  extra.id = "typebridge.generator.typescript.zzz-extra-resource";
  value.code_resources.push(extra);
  replaceProjection(directory, value);
});
rejectSuccessor("duplicate_resource", (directory) => {
  const value = projection(directory);
  value.code_resources.push(structuredClone(value.code_resources[0]));
  replaceProjection(directory, value);
});
rejectSuccessor("reordered_resource", (directory) => {
  const value = projection(directory);
  value.code_resources.reverse();
  replaceProjection(directory, value);
});
rejectSuccessor("forged_resource", (directory) => {
  const value = projection(directory);
  value.code_resources[0].content_fingerprint.digest = "0".repeat(64);
  replaceProjection(directory, value);
});
rejectSuccessor("stale_handler", (directory) => {
  const value = projection(directory);
  value.generator_handlers[0].version = 1;
  replaceProjection(directory, value);
});
rejectSuccessor("foreign_target", (directory) => {
  const value = projection(directory);
  value.target = "python";
  replaceProjection(directory, value);
});
rejectSuccessor("malformed_semantic_fingerprint", (directory) => {
  replaceExport(
    directory,
    "dist/schema.js",
    SEMANTIC_FINGERPRINT_PREFIX,
    "{}",
  );
});
rejectSuccessor("malformed_projection_fingerprint", (directory) => {
  replaceExport(
    directory,
    "dist/schema.js",
    PROJECTION_FINGERPRINT_PREFIX,
    "{}",
  );
});

console.log("generated TypeScript authority and projection rejection acceptance passed");
